lib/cli/interactive-setup/aws-credentials.js

```diff
@@ -86,33 +178,141 @@ const steps = {
+  },
+  // TODO: TESTS
+  linkProviderToService: async (app, service, orgName, providerUid) => {
+    const sdk = await getSdkInstance(orgName);
+    const linkType = 'service';
+    const linkUid = getProviderLinkUid(app, service);
+    const orgUid = await getOrgUidByName(orgName);
+    return sdk.createProviderLink(orgUid, linkType, linkUid, providerUid);
+  },
 };
 
 module.exports = {
-  async isApplicable({ configuration }) {
+  async isApplicable(context) {
+    const { configuration } = context;
     if (
       _.get(configuration, 'provider') !== 'aws' &&
       _.get(configuration, 'provider.name') !== 'aws'
     ) {
       return false;
     }
     if (new AWS.S3().config.credentials) return false;
-    return !(await awsCredentials.resolveFileProfiles()).size;
+    if ((await awsCredentials.resolveFileProfiles()).size) return false;
+
+    // TODO: TESTS
+    const orgName = _.get(configuration, 'org');
+    if (orgName) {
+      const providers = await getProviders(orgName);
+      const hasDefaultProvider = providers.some((provider) => provider.isDefault);
+
+      if (hasDefaultProvider) return false;
+    }
+
+    return true;
   },
   async run(data) {
-    process.stdout.write(
-      'No AWS credentials were found on your computer, ' +
-        'you need these to host your application.\n\n'
-    );
-
     return module.exports.runSteps(data);
   },
   steps,
   runSteps: async (context) => {
-    if (!(await steps.shouldSetupAwsCredentials())) return;
-    await steps.ensureAwsAccount();
+    const providers = await getProviders(context.configuration.org);
+    const credentialsSetupChoiceAnswer = await credentialsSetupChoice(providers);
+
+    if (credentialsSetupChoiceAnswer === CREDENTIALS_SETUP_CHOICE.CREATE_PROVIDER) {
+      try {
+        await steps.handleProviderCreation(context);
+      } catch (err) {
+        throw new ServerlessError(
+          `Could not handle setup of a new provider in Dashboard: ${err.message}`,
+          'CANNOT_SETUP_NEW_PROVIDER'
+        );
+      }
+      return;
+      // TODO: CONSIDER IF WE SHOULD CREATE DIRECT LINK for default provider
+    } else if (credentialsSetupChoiceAnswer === CREDENTIALS_SETUP_CHOICE.SKIP) {
+      // TODO: Add prompt that you can set it up later - link to docs
+      return;
+    } else if (credentialsSetupChoiceAnswer === CREDENTIALS_SETUP_CHOICE.LOCAL) {
+      await steps.ensureAwsAccount();
+      await steps.ensureAwsCredentials(context);
+      await steps.inputAwsCredentials();
+      return;
+    }
+    // Otherwise user selected an existing provider
+
+    try {
+      await steps.linkProviderToService(
+        context.configuration.app,
+        context.configuration.service,
+        context.configuration.org,
+        credentialsSetupChoiceAnswer
+      );
+    } catch (err) {
+      throw new ServerlessError(
+        `Could not link the selected provider to your service: ${err.message}`,
+        'CANNOT_LINK_EXISTING_PROVIDER'
+      );
```

Comment:
Similarly here, it feels too broad, doesn't feel as good design to just catch all errors into this slot (we risk hiding programmer errors)

Reply:
So, do you suggest to go with specific error checks inside `linkProviderToService` and explicit messages about what went wrong, even if that is becoming more "technical"? I wasn't sure how deep to go because I feel like "ServerlessError" is mostly for some user configuration error, while here we can have different errors that don't really are user-related such as e.g. broken connection, unavailable API, etc

Reply:
When I look into implementation of `linkProviderToService` I see that hiding all errors that can happen there and translating them to `Could not link the selected provider to your service` can be error prone.
Technically it's like sweeping under carpet anything bad that may happen, also any programmer error we may accidentally introduce in this file.

This wrap would be ok, if error handling would be as:

- If `ServerlessError` pass through (if `ServerlessError` was thrown it means it already has user facing message we should show)
- If error comes directly from SDK. Ideally would be to detect if it's a _service_ error, or _user_ error.
   - _service_ error should be hidden behind something `Sorry dashboard service is currently inaccessible, please try again later` and then also it'll be good to not throw, but skip this step and follow.
    - _user_ error (so error in how SDK was used) should be propagated as programmer error (so pass through, dirty crash, eventually we may tweak message to give more context, so when we see this in report, we have better idea where it happened)
- Any other error, pass through, it's a programmer error (bug) we should not hide.

Of course it's easy to state above, while I know that our SDK is poor at communicating errors. So if at this point we cannot easily distinguish between _system_ and _user_ errors as coming from SDK, what I propose is to ensure to expose each SDK call error with `Sorry dashboard service is currently inaccessible, please try again later`, and ensure any other errors do not fall into that.

I think we can achieve that with:
- Internally ensure that each SDK call fail results with distinguishable error (so we know it's an SDK fail)
- In this wrap, handle just SDK fails (with `Sorry dashboard service is currently inaccessible, please try again later` message and gentle return) and pass other errors through